examples/jupyter/tutorials/tutorial4-python-elemrestriction.ipynb

```diff
@@ -0,0 +1,511 @@
+   "metadata": {},
+   "source": [
+    "Then move to the libCEED folder"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "cd libCEED"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "And compile the base library by running"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "! make"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Build libCEED for Python by running"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "python setup.py build install"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Or without superuser permissions"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "python setup.py build install --user"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## CeedElemRestriction\n",
+    "\n",
+    "Here we show some basic examples to illustrate the `CeedElemRestriction` class. In libCEED, a `CeedElemRestriction` groups the degrees of freedom (dofs) of the local vector according to the different elements they belong to (see [the API documentation](https://libceed.readthedocs.io/en/latest/libCEEDapi.html#finite-element-operator-decomposition)).\n",
+    "\n",
+    "Here we illustrate the simple creation and application of a `CeedElemRestriction`, with user provided dof indices"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "[10. 11. 11. 12. 12. 13.]\n"
+     ]
+    }
+   ],
+   "source": [
+    "import libceed\n",
+    "import numpy as np\n",
+    "\n",
+    "# In this 1D example, the dofs are indexed as\n",
+    "#\n",
+    "# Restriction input:\n",
+    "#  x --  x --  x --  x\n",
+    "# 10 -- 11 -- 12 -- 13\n",
+    "#\n",
+    "# Restriction output:\n",
+    "#  x --  x |  x --  x | x --  x\n",
+    "# 10 -- 11 | 11 -- 12 | 12 -- 13\n",
+    "\n",
+    "ceed = libceed.Ceed()\n",
+    "\n",
+    "ne = 3\n",
+    "\n",
+    "x = ceed.Vector(ne+1)\n",
+    "a = np.arange(10, 10 + ne+1, dtype=\"float64\")\n",
+    "x.set_array(a, cmode=libceed.USE_POINTER)\n",
+    "\n",
+    "ind = np.zeros(2*ne, dtype=\"int32\")\n",
+    "for i in range(ne):\n",
+    "  ind[2*i+0] = i\n",
+    "  ind[2*i+1] = i+1\n",
+    "    \n",
+    "r = ceed.ElemRestriction(ne, 2, ne+1, 1, ind, cmode=libceed.USE_POINTER)\n",
+    "\n",
+    "y = ceed.Vector(2*ne)\n",
+    "y.set_value(0)\n",
+    "\n",
+    "r.apply(x, y)\n",
+    "\n",
+    "y_array = y.get_array_read()\n",
+    "print(y_array)\n",
+    "y.restore_array_read()"
+   ]
```

Comment:
I agree that since we support the `print( )` for restrictions we could use those as well. But I was refraining from doing that because the output in that case would look like

```
CeedVector length 6
  10.000000
  11.000000
  11.000000
  12.000000
  12.000000
  13.000000
```
showing the C-style floating point format, while by using the print of the numpy.array it displays the integer values nicely. And since these are indices in this particular example, I thought it would confuse learners to see doubles instead of integers.

Reply:
I have changed it for all other print statements (bases, qfunctions etc.) except for restrictions, for which I think the integer display is nicer. Thanks again.